igor_src/main.cpp

```diff
@@ -1876,8 +1954,193 @@ int main(int argc , char* argv[]){
+										return (dels <= cdr3_v_read_anch); 
+							}
+						);
+					cdr3_v_read_anch = cdr3_v_read_anch + dels_correction; // ins_size before the cdr3 a
+
+
+					///////////////////////////////////
+					// Get only one alignment for J  //
+					///////////////////////////////////
+					vector<Alignment_data> Vec_J_Alignment_data = (sorted_alignments[seq_index].second )[J_gene];
+					
+					// If no J alignment was found then write an empty CDR3
+					if ( Vec_J_Alignment_data.size() == 0 ){
+						cout << "seq_index = " << seq_index << ", Vec_V_Alignment_data.size() = " << Vec_J_Alignment_data.size() <<endl;
+						ofileIndexedCDR3 << seq_index << ";";
+						ofileIndexedCDR3 << cdr3_v_read_anch  << ";;;;"<<endl;
+						continue;
+					}
+					
+					Alignment_data j_alig = Vec_J_Alignment_data.front(); //at(0);
+					
+					j_gene_name = j_alig.gene_name;
+					j_gene_str  = UMap_j_genomic[j_alig.gene_name];
+					j_ins_size  = distance(j_alig.insertions.begin(), j_alig.insertions.end() );
+					j_dels_size = distance(j_alig.deletions .begin(), j_alig.deletions .end() );
+
+					// Get the anchor from map and correct them.
+					cdr3_j_gene_anch = UMap_j_CDR3_anchors[j_alig.gene_name];
+					int j_ins_correction = std::count_if(
+							j_alig.insertions.begin(), j_alig.insertions.end(), 
+							// Lambda function for condition
+							[cdr3_j_gene_anch](int inss){ 
+										return (inss <= cdr3_j_gene_anch); 
+							}
+					);
+					cdr3_j_gene_anch = cdr3_j_gene_anch + j_ins_correction;
+
+					int cdr3_j_read_anch = cdr3_j_gene_anch + j_alig.offset;
+					int j_dels_correction = std::count_if(
+							j_alig.deletions.begin(), j_alig.deletions.end(), 
+							// Lambda function for condition
+							[cdr3_j_read_anch](int dels){ 
+										return (dels <= cdr3_j_read_anch); 
+							}
+						);
+
+					// In order to get the phelanine or triptophan in the sequence +3.
+					cdr3_j_read_anch = cdr3_j_read_anch + j_dels_correction + 3; // ins_size before the cdr3 anchor
+					cout << seq_str.size() << ", "<< cdr3_v_read_anch << ", " << cdr3_j_read_anch << endl;	
+					// --------------- Print in file if anchor is coherent ---------------- //
+					ofileIndexedCDR3 << seq_index << ";";
+					ofileIndexedCDR3 << cdr3_v_read_anch  << ";";
+					ofileIndexedCDR3 << cdr3_j_read_anch  << ";" ;
+					// Anchor is inside proposed alignment.
+					if ( ( cdr3_v_read_anch <= seq_str.size() ) and
+							 ( cdr3_j_read_anch <= seq_str.size() ) and
+							 ( cdr3_v_read_anch <= cdr3_j_read_anch )	){
+						cout << seq_str<<", " <<endl;
+						string strCDR3 = seq_str.substr(cdr3_v_read_anch, cdr3_j_read_anch - cdr3_v_read_anch);
+//						ofileIndexedCDR3 << cdr3_j_read_anch  << ";";
+						ofileIndexedCDR3 << strCDR3 << ";";
+						cout << strCDR3 << ";";
+						ofileIndexedCDR3 << translate(strCDR3);
+					} 
+
+					ofileIndexedCDR3 << endl;
+				}// end for indexed_seqlist
+				ofileIndexedCDR3.close();
+			}
+		}//end align
```

Comment:
This is a complicated procedure (more than 100 lines of code), this should make for a standalone function for readability and re-usability. Maybe the Aligner file would be a better place for such a function?

Reply:
Yep I know and I absolutely agree with you, and at the time I made it just to work. I can change it now but I need to think a little bit more how. Because when I started with that my idea was to create a struct or class called CDR3SeqData and made a vector with this struct, quite similar to the vector indexed_seqlist, and use a functor class called GeneFeatures to encapsulate the data of gene anchors files into maps UMap_v_CDR3_anchors and UMap_j_CDR3_anchors (or other possible anchors). Then pass as arguments a pair of the indexeded_seqlist or just the ID, the corresponding V and J Alignment_data structures and return the wanted gene feature as CDR3SeqData. And finally something like GeneFeatures::save_CDR3(strFilename). But as I told you I need to think about adding things in a consistent way.

Reply:
Thanks for the work, it's much more compact and clear now!